### PR TITLE
Skip shards when querying constant keyword fields

### LINKED ISSUE #95541: Avoid refreshing search-idle shards that don't yield results after query rewrite
Many search requests have the following structure:

```
...
"query": {
      "bool": {
        "must": [],
        "filter": [
          {
            "match_phrase": {
              "data_stream.dataset": "kubernetes.container"
            }
          },
          {
            "range": {
              "@timestamp": {
                "format": "strict_date_optional_time",
                "gte": "...",
                "lte": "...."
              }
            }
          }
        ],
        "should": [],
        "must_not": []
      }
    }
...
```

The index pattern matches (`metrics-*`) matches all metric data streams, but the `match_phrase` query on the `data_stream.dataset` field, which is a constant keyword field, only matches with one specific data stream.

Before query rewriting either in the `can_match` or `query` phases, shards that are search-idle get refreshed. This increases the query time significantly. Many o11y use cases rely on the default refresh behaviour. Which is the schedule a refresh every second when a shard is search active and don't schedule any refreshes when a shard is search-idle, this to favour indexing performance.

The refresh that occurs before the query rewrite should not occur on shards that don't match with the required filter clause on the `data_stream.dataset` constant keyword field. That is the goal of this issue..

### PR DESCRIPTION
**Pull request by salvatore-campagna**
When a query like a match phrase query or a wildcard query targets a constant keyword field we can skip the query execution on shards where the query builder is rewritten to a MatchNoneQueryBuilder.

This allows us to save time and resources executing queries on nodes that do not return any result. Constant keyword fields allow rewriting a MatchPhraseQueryBuider to TermsQueryBuilder which, in turn, is rewritten into a MatchNoneQueryBuilder in case a constant keyword field does not match the value specified by the query.

A match phrase query is often executed on constant keyword fields by different Elasticsearch integrations which target multiple indices by means of index patterns or data streams. In a real world scenario it is likely that the index pattern or data stream includes tens or hundreds of backing indices each of them with multiple shards involved. As a result, skipping shards in such scenario might result in better query performance and better cluster resource usage.

Note, anyway, that currently, execution of the pre-filter and the corresponding "can match" phase depends on the overall number of shards involved and on whether there is at least one of them returning a non-empty result.

We do the rewrite operation on the data node in the so called "can match" phase, taking advantage of the fact that, at that moment, we can access the index mapping and extract information about constant keyword fields and their value. Doing this on the coordinator node is not possible due to the unavailability of the index mapping.

Resolves #95541

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -1524,6 +1524,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
...
+                    request.shardId(),
+                    request.shardRequestIndex(),
+                    request.getClusterAlias(),
+                    request.getRuntimeMappings().isEmpty() == false
+                ).get()
+            );
+            if (rewritten.query() instanceof MatchNoneQueryBuilder) {
+                return new CanMatchShardResponse(false, null);
+            }
+        }
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1542]: The `MappingAwareRewriteContext` doesn't seem to be needed. Why not creating a new `SearchExecutionContext` when the index service is [available](https://github.com/elastic/elasticsearch/pull/96161/files#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fL1567).
You should also reuse `queryStillMatchesAfterRewrite(request, context)` to ensure that all cases are checked (alias filter and global aggregations).

**Review by jimczi — COMMENTED** [review]
I looked at the PR because I thought it would be difficult to achieve but looks like we already handle the case where the `SearchExecutionContext` doesn't have a searcher. Nice one!
I left one comment to simplify further.
I also wonder why the optim is disabled when runtime fields are added in the request? 
What are we protecting against?

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1542]: > If we use SearchExecutionContext then also range queries (and maybe others?) could be resolved to MatchNoneQueryBuilder instance.

The initial change already used `SearchExecutionContext` but it sets the searcher to `null`. That's why I said that the special context is not needed. There's some logic to handle `SearchExecutionContext` with a null searcher, we should add some javadocs to explain the use case (rewriting with the mapping only) but it works.

Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1542]: Right, I see this now. Yes, if we pass a null searcher then this should work. 👍

Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1542]: I had to add an additional null check in `isFieldWithinQuery`. That logic uses the reader which might be null in this case and triggers a NPE at some point.

File: server/src/test/java/org/elasticsearch/search/SearchServiceTests.java [hunk 2]
```diff
@@ -1027,13 +1027,13 @@ public void testCanMatch() throws Exception {
 
         CountDownLatch latch = new CountDownLatch(1);
         SearchShardTask task = new SearchShardTask(123L, "", "", "", null, Collections.emptyMap());
-        assertEquals(8, numWrapInvocations.get());
+        assertEquals(6, numWrapInvocations.get());
```
Comment (salvatore-campagna) [inline server/src/test/java/org/elasticsearch/search/SearchServiceTests.java:1030]: I had to change these values because we early exit due to shard filtering and the counter is not incremented.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 3]
```diff
@@ -1532,6 +1532,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
 
     private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean checkRefreshPending) throws IOException {
         assert request.searchType() == SearchType.QUERY_THEN_FETCH : "unexpected search type: " + request.searchType();
+        if (request.source() != null && (request.source().query() instanceof MatchNoneQueryBuilder) == false) {
+            final SearchExecutionContext searchExecutionContext = new SearchExecutionContext(
+                indicesService.indexServiceSafe(request.shardId().getIndex())
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1537]: We should reuse the `indexService` below, using `indexServiceSafe` might not work for PIT if the index was deleted after the creation. Can you move this code [here](https://github.com/elastic/elasticsearch/pull/96161/files#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fL1575)?

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 4]
```diff
@@ -1532,6 +1532,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
...
+                    .newSearchExecutionContext(
+                        request.shardId().getId(),
+                        request.shardRequestIndex(),
+                        null,
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1541]: Using `null` here is okish but we should update the javadoc in [SearchExecutionContext](https://github.com/elastic/elasticsearch/blob/29ecff1802ca31fcb210bd54164723eeab94b646/server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java#L685)?

**Review by jimczi — COMMENTED** [review]
I left some additional comments. Thanks for the simplification!

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 3 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1537]: The idea of doing it at the beginning was to do a early exit and avoid doing everything that happens after...

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1537]: These operations are cheap (no IO) so it shouldn't be an issue and we want to handle all cases (PIT and normal search)?

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 5]
```diff
@@ -1580,6 +1564,21 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
...
+                            null,
+                            request::nowInMillis,
+                            request.getClusterAlias(),
+                            request.getRuntimeMappings()
+                        )
+                    );
+                    if (queryStillMatchesAfterRewrite(request, searchExecutionContext) == false) {
+                        return new CanMatchShardResponse(false, null);
+                    }
+                }
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1581]: As I said in my previous comment, it's safe to do that for all cases (PIT and normal searcher). Please move the logic [here](https://github.com/elastic/elasticsearch/pull/96161/files/0de29feb016ae7516603a6198d427b03ebc35db6..1ce266bd43af65169d1d2b1cbf7def2a8230c12b#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fR1590).
Can you also add a comment explaining why rewriting without a searcher allows to ignore hasRefreshPending?

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 3 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: I think we need to do this rewrite before we invoked `IndexShard#acquireSearcher(...)` since this marks the shard as search active again (and would allow scheduled refreshes to really be executed). So maybe we should keep the rewrite in the place it was?

Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: If we do it late shards go into 'search active state'. `acquireSearcher` does `markSearcherAccessed `. WHich makes the shard search active which will cause waiting for refresh...which we want to avoid.

Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchService.java]: I see and the PIT case sets the refresh pending to false anyway so no need to change that. Thanks for spotting.

File: server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleTests.java [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.IntToLongFunction;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertNoSearchHits;
+import static org.hamcrest.Matchers.equalTo;
+
+public class SearchIdleTests extends ESSingleNodeTestCase {
```
Comment (martijnvg) [inline server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleTests.java:48]: Why was this test renamed?
I think all test cases need to have the suffix `IT` in `internalClusterTest`?

File: server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java [hunk 7]
```diff
@@ -681,7 +681,10 @@ public IndexReader getIndexReader() {
     }
 
     /** Return the current {@link IndexSearcher}, or {@code null} if no index reader is available,
-     *  for instance if this rewrite context is used to index queries (percolation). */
+     * which happens in two cases:
+     * 1. if this rewrite context is used to index queries (percolation)
+     * 2. if we use mapping information to skip shards while doing shards pre-filtering in the 'can match' phase
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java:686]: Basically when a `SearchExecutionContext` is used with no indexReader.

**Review by martijnvg — COMMENTED** [review]
In general this looks good. Left a few minor comments.

File: server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java [hunk 8]
```diff
@@ -692,6 +692,10 @@ public Relation isFieldWithinQuery(
             DateMathParser dateParser,
             QueryRewriteContext context
         ) throws IOException {
+            if (reader == null) {
+                // NOTE: in case we rewrite a query before we have a reader available
+                return Relation.INTERSECTS;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:697]: is this uncovering an existing bug that may be triggered by the percolator, or only caused by doing a rewrite before acquiring a searcher?

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 9]
```diff
@@ -1596,6 +1608,37 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
...
+                null,
+                request::nowInMillis,
+                request.getClusterAlias(),
+                request.getRuntimeMappings()
+            )
+        );
+        if (queryStillMatchesAfterRewrite(request, searchExecutionContext) == false) {
+            return new CanMatchShardResponse(false, null);
+        }
+        return null;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1639]: Perhaps this should return a boolean. If true we will go ahead, otherwise we will return a can match shard response with the flag set to false?

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 10]
```diff
@@ -1545,13 +1545,21 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
                     readerContext = findReaderContext(request.readerId(), request);
                     releasable = readerContext.markAsUsed(getKeepAlive(request));
                     indexService = readerContext.indexService();
+                    final CanMatchShardResponse canMatchAfterRewrite = canMatchAfterRewriteWithoutSearcher(request, indexService);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1548]: I am slightly uneasy with calling rewrite yet another time. We do this because acquireSearcher will wake up the shard from idle state and there is no way around it, meaning if we get a searcher we do need to refresh the shard which is precisely what we are trying to avoid?

**Review by javanna — COMMENTED** [review]
left a couple of comments

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 11]
```diff
@@ -1596,6 +1608,37 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
...
+     */
+    private static CanMatchShardResponse canMatchAfterRewriteWithoutSearcher(
+        final ShardSearchRequest request,
+        final IndexService indexService
+    ) throws IOException {
+        final SearchExecutionContext searchExecutionContext = new SearchExecutionContext(
+            indexService.newSearchExecutionContext(
+                request.shardId().getId(),
+                request.shardRequestIndex(),
+                null,
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1630]: I wonder if passing a null searcher warrants a cleanup in a followup. I know it's a thing that may already happen but I am not sure we should build upon that further. The rewrite needs a `QueryRewriteContext`, which though will call `convertToSearchExecutionContext` to call `getFieldType` against it. Maybe the query rewrite context could expose  `getFieldType` that would work when mappings are available (in the data node that has the index), to cover for situations where we want to rewrite purely based on mappings like in this case. I see though that `ConstantFieldType#matches` requires the `SearchExecutionContext` as an argument so the methods needed are more than just `getFieldType` and the change I am suggesting is more involving than it may seem... @romseygeek double checking if you have ideas

File: server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java (hunk 8 above)
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java:697]: No bug, just doing a rewrite before we have a searcher available.

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 11 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1625]: I think it's just `getFieldType` and `getIndexSettings()`, both of which can be happily moved into `QueryRewriteContext` I think?  Ideally we could move a couple of more methods in there and remove the abomination that is `convertToSearchExecutionContext()`....

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1625]: And `indexNamedMatcher`.  And actually we can move `getWriteableRegistry` and `nowInMillis` into `SearchExecutionContext` as they are never used for rewriting, which nearly balances things out.

Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1625]: @romseygeek so if I understand correctly the idea is to move all of that stuff into `QueryRewriteContext` so we don't need a `SearchExecutionContext` at all?

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1625]: Yes, precisely.  I'm working on a PR now...

**Comment by salvatore-campagna** [comment]
@romseygeek @javanna is it ok to do the refactoring (moving things in `QueryRewriteContext` and `SearchExecutionContext` in a followup PR?

**Comment by salvatore-campagna** [comment]
@elasticsearchmachine test this please

**Comment by salvatore-campagna** [comment]
@martijnvg @javanna @romseygeek I need at leat one approval to proceed with merging.

File: server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java [hunk 12]
```diff
@@ -223,4 +231,216 @@ public void testSearchIdleStats() throws InterruptedException {
...
+                        .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, activeIndexShardsCount)
+                        .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), -1)
+                        .put(IndexSettings.MODE.getKey(), IndexMode.TIME_SERIES)
+                        .put(IndexMetadata.INDEX_ROUTING_PATH.getKey(), "routing_field")
+                        .put(IndexSettings.TIME_SERIES_START_TIME.getKey(), "2021-05-12T00:00:00.000Z")
+                        .put(IndexSettings.TIME_SERIES_END_TIME.getKey(), "2021-05-13T23:59:59.999Z")
+                )
+                .get()
+        );
+        assertAcked(
```
Comment (martijnvg) [inline server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java:274]: Can these mappings be added with the create index call?

File: server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java [hunk 13]
```diff
@@ -223,4 +231,216 @@ public void testSearchIdleStats() throws InterruptedException {
...
+        assertEquals(1, searchResponse.getHits().getHits().length);
+
+        // THEN
+        final IndicesStatsResponse idleIndexStatsAfter = client().admin().indices().prepareStats(idleIndex).get();
+        Arrays.stream(idleIndexStatsAfter.getShards()).forEach(shardStats -> assertTrue(shardStats.isSearchIdle()));
+
+        final IndicesStatsResponse activeIndexStatsAfter = client().admin().indices().prepareStats(activeIndex).get();
+        Arrays.stream(activeIndexStatsAfter.getShards()).forEach(shardStats -> assertFalse(shardStats.isSearchIdle()));
+
+        assertRefreshStats(idleIndexStatsBefore, idleIndexStatsAfter);
```
Comment (martijnvg) [inline server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java:355]: I think we should also have an `assertActiveRefreshStats(...)`? This ensures that total of after is higher than before.

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 14]
```diff
@@ -1596,6 +1605,32 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
...
     }
 
+    /**
+     * This method tries to rewrite a query without using a searcher (before we get a valid one). It takes advantage of the fact that
+     * we can skip some shards in the query phase because we have enough information in the index mapping to decide the 'can match'
+     * outcome. One such example are queries like {@link org.elasticsearch.index.query.MatchPhraseQueryBuilder} queries which are rewritten
+     * to {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} after being rewritten to
+     * {@link org.elasticsearch.index.query.TermQueryBuilder}. Other queries have a similar pattern, especially when using
+     * {@link org.elasticsearch.index.mapper.ConstantFieldType} or unmapped fields, they all rewrite to
+     * {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} (if conditions are met). This allows us to avoid extra work other than
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1615]: Maybe this text of this example can be simplified? Something like: `On such example is a term based query against a constant keyword field. This queries can rewrite themselves to a `MatchNoneQueryBuilder`.

File: x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+        // THEN
+        final IndicesStatsResponse afterStatsResponse = client().admin().indices().prepareStats("test*").get();
+        final List<ShardStats> activeShards = Arrays.stream(afterStatsResponse.getShards())
+            .filter(shardStats -> shardStats.isSearchIdle() == false)
+            .toList();
+        // NOTE: we need an empty result from at least one shard
+        assertEquals(1, activeShards.size());
+
+        assertRefreshStats(statsResponse, afterStatsResponse);
```
Comment (martijnvg) [inline x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:89]: Maybe as in the other test case?

File: x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+            .setQuery(new MatchPhraseQueryBuilder(field, value))
+            .setPreFilterShardSize(preFilterShardSize)
+            .get();
+    }
+
+    private static void assertRefreshStats(final IndicesStatsResponse before, final IndicesStatsResponse after) {
+        final List<RefreshStats> refreshStatsBefore = Arrays.stream(before.getShards()).map(x -> x.getStats().refresh).toList();
+        final List<RefreshStats> refreshStatsAfter = Arrays.stream(after.getShards()).map(x -> x.getStats().refresh).toList();
+        assertTrue(
+            refreshStatsBefore.size() == refreshStatsAfter.size()
```
Comment (martijnvg) [inline x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:270]: Maybe as in the other test case?

File: x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+        // WHEN
+        final SearchResponse searchResponse = search("test*", "constant_keyword", "constant_value2", 5);
+        assertEquals(RestStatus.OK, searchResponse.status());
+        assertEquals(idleIndexShardsCount, searchResponse.getSkippedShards());
+        assertEquals(0, searchResponse.getFailedShards());
+        Arrays.stream(searchResponse.getHits().getHits()).forEach(searchHit -> assertEquals("test2", searchHit.getIndex()));
+
+        // THEN
+        final IndicesStatsResponse idleIndexStatsAfter = client().admin().indices().prepareStats(idleIndex).get();
+        Arrays.stream(idleIndexStatsAfter.getShards()).forEach(shardStats -> assertTrue(shardStats.isSearchIdle()));
```
Comment (martijnvg) [inline x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:137]: Maybe assert that `idleIndexStatsAfter.getShards()` has elements or isn't empty?

Comment (martijnvg) [reply x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:137]: Maybe turn this then into a static helper method? So that we do this for other similar assertions too.

**Review by martijnvg — APPROVED** [review]
Left a few minor comment. Otherwise LGTM.

File: server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java (hunk 13 above)
Comment (salvatore-campagna) [reply server/src/internalClusterTest/java/org/elasticsearch/index/shard/SearchIdleIT.java:355]: I am not sure we can match results shard by shard...

File: x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+
+    private void assertActiveShardsRefreshStats(final IndicesStatsResponse before, final IndicesStatsResponse after) {
+        assertNotEquals(0, before.getShards().length);
+        assertNotEquals(0, after.getShards().length);
+        final Map<ShardRouting, ShardStats> afterStatsAsMap = after.asMap();
+        for (var beforeShard : before.getShards()) {
+            final RefreshStats beforeRefresh = beforeShard.getStats().refresh;
+            assertTrue(afterStatsAsMap.containsKey(beforeShard.getShardRouting()));
+            final RefreshStats afterRefresh = afterStatsAsMap.get(beforeShard.getShardRouting()).getStats().refresh;
+            assertTrue(afterRefresh.getTotal() >= beforeRefresh.getTotal());
```
Comment (salvatore-campagna) [inline x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:311]: Here we should need just `>` (not equals) because a refresh should have happened in the meanwhile after executing the search query. Anyway, if I remove the equal the testing refresh stats on the active index fails because the values are the sane before and after. I am trying to figure out why.

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 11 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1622]: I'd advocate for the needed refactoring to be a prerequisite of the PR. I think it does not make a lot of sense to me to build on the search context with null searcher if we want to rather rely on a cleaner approach that relies on the query rewrite context. Otherwise we add in tech debt that we are going to have to remove straight-away or will stay around for much longer if we don't get to it? What do you folks think?

**Comment by javanna** [comment]
> is it ok to do the refactoring (moving things in QueryRewriteContext and SearchExecutionContext) in a followup PR?

I commented above. It's not ideal, I think. I would go that way only if the suggested refactoring takes long or we don't plan to work on it at the moment, which does not seem to be the case. @salvatore-campagna how would you feel about working towards the direction that Alan pointed you to?

**Comment by romseygeek** [comment]
Looking again at some of this, would it be possible to reuse the `CoordinatorRewriteContext` for anything here?  The API could be something like:

```
MyGenericQueryBuilder.java

protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
    CoordinatorRewriteContext crc = queryRewriteContext.convertToCoordinatorRewriteContext();
    if (crc != null) {
        // we're on the coordinator, do something simple
    }
    SearchExecutionContext sec = queryRewriteContext.convertToSearchExecutionContext();
    if (sec != null) {
        // on the shard, ready to execute the query, do the final rewrite
    }
    // on the shard, in the can match phase, do something with mappings and/or metadata but no searcher
}
```

**Comment by salvatore-campagna** [comment]
> 

Yes I can work on that but I think this is out of the scope of this PR and makes it way too large (also to review). SO I suggest to do it in a followup PR, and I can do it. Ok?

**Comment by javanna** [comment]
> Yes I can work on that but I think this is out of the scope of this PR and makes it way too large (also to review).

Thanks! I was not suggesting to make all needed changes as part of this PR. I would suggest opening separate prerequisite PRs that allow us to make the desired change (skipping shards when querying constant keyword fields) without relying on a null searcher in search context. Getting this PR in first looks to me like adding tech debt that we are forced to tackle straightaway, and unless there is urgency to get some fix in asap, I would advocate for taking the clean approach if possible. I am assuming the desired change will still make 8.9.

**Comment by salvatore-campagna** [comment]
> > Yes I can work on that but I think this is out of the scope of this PR and makes it way too large (also to review).
> 
> Thanks! I was not suggesting to make all needed changes as part of this PR. I would suggest opening separate prerequisite PRs that allow us to make the desired change (skipping shards when querying constant keyword fields) without relying on a null searcher in search context. Getting this PR in first looks to me like adding tech debt that we are forced to tackle straightaway, and unless there is urgency to get some fix in asap, I would advocate for taking the clean approach if possible. I am assuming the desired change will still make 8.9.

Thanks Luca,

I will create a new issue where I describe what we need to do for the refactoring and keep this PR open. I will work on it.

**Comment by salvatore-campagna** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1  please.

**Comment by salvatore-campagna** [comment]
@javanna @romseygeek @martijnvg I adjusted this PR after merging #96353

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 19]
```diff
@@ -1595,6 +1604,23 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
...
 
+    /**
+     * This method tries to rewrite a query without using a {@link SearchExecutionContext}. It takes advantage of the fact that
+     * we can skip some shards in the query phase because we have enough information in the index mapping to decide the 'can match'
+     * outcome. One such example is a term based query against a constant keyword field. This queries can rewrite themselves to a
+     * {@link MatchNoneQueryBuilder}. This allows us to avoid extra work other than
+     * making the shard search active and waiting for refreshes. As a result, we only wait for refreshes to happen on shards that have
+     * actual data.
+     */
+    private static boolean canMatchAfterRewrite(final ShardSearchRequest request, final IndexService indexService) throws IOException {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1615]: Maybe reword: `This allows us to avoid extra work other than making the shard search active and waiting for refreshes.` to `This allows us to avoid extra work for example making the shard search active and waiting for refreshes.`?

**Review by martijnvg — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 20]
```diff
@@ -661,6 +668,46 @@ public SearchExecutionContext newSearchExecutionContext(
         );
     }
 
+    /**
+     * Creates a new {@link QueryRewriteContext}.
+     * This class is used to rewrite queries before we are able to get a valid {@link SearchExecutionContext} and
+     * allows us to anticipate rewriting queries before the query is executed on the data node. Not using a full
+     * SearhcExecutionContext allows us to save on query latency and IO operations.
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/IndexService.java:675]: SearchExecutionContext is mistyped. Shall we also clarify that this is for cases where access to the index is not required as we can make decisions based on mappings alone? Also, expand on what saves query latency and IO operations? e.g. not pulling a searcher removes the cost of pulling a searcher as well as the associated cost of refreshing idle shards.

File: server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java (hunk 7 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java:627]: is the second still true given the recent improvements you made? Are there places where we do mappings based can match with a null searcher?

File: x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java [hunk 21]
```diff
@@ -0,0 +1,319 @@
...
+        final List<RefreshStats> refreshStatsBefore = Arrays.stream(before.getShards())
+            .map(shardStats -> shardStats.getStats().refresh)
+            .toList();
+        final List<RefreshStats> refreshStatsAfter = Arrays.stream(after.getShards())
+            .map(shardStats -> shardStats.getStats().refresh)
+            .toList();
+        assertEquals(refreshStatsBefore.size(), refreshStatsAfter.size());
+        assertTrue(refreshStatsAfter.containsAll(refreshStatsBefore));
+        assertTrue(refreshStatsBefore.containsAll(refreshStatsAfter));
+    }
```
Comment (javanna) [inline x-pack/plugin/mapper-constant-keyword/src/test/java/org/elasticsearch/xpack/constantkeyword/mapper/SearchIdleTests.java:318]: It is odd that we don't have unit tests, yet I see that SearchServiceTests that already exists is also an `ESSingleNodeTestCase`. You may want to simplify your test taking inspiration from it, in that it pulls the search service from the node and calls canMatch directly against it.

**Review by javanna — APPROVED** [review]
I left a couple of comments, especially one on testing, LGTM otherwise.
